### DISCUSSION TITLE
io_uring_register_buffers() failed: Cannot allocate memory

### DISCUSSION THREAD

**Discussion by jmillan** [axboe/liburing#1021] [Q&A]
liburing version: 2.4

I can read in the method [man](https://manpages.debian.org/unstable/liburing-dev/io_uring_register.2.en.html#IORING_REGISTER_BUFFERS):

```txt
The buffers associated with the iovecs will be locked in memory and charged against the user's RLIMIT_MEMLOCK resource limit
```

I'm passing `io_uring_register_buffers()` an array of 4096 `iovec`s of 1500B len each. (~=5MB in total). Such memory was already allocated by the process.

I'm having this issue when running the [GitHub actions for my project](https://github.com/versatica/mediasoup/pull/1273), and it's happening just for Ubuntu 22.04 version, not older ones. I would like to understand the reason for this error.

Does it mean that there are no 5MB for the process left? Would it be fixed by increasing the RLIMIT_MEMLOCK[*]?

[*]: I'm asking and not just doing it because the images are out of my control.

Thanks.

**Comment by axboe** [comment]
Sounds like it's RLIMIT_MEMLOCK indeed. What does ```ulimit -l``` say for your user? Some distros will have this set very low, like 64k or similar.

A separate question is why you are registering 1500b buffers, that sounds like networking to me. Registered buffers will work with networking, but they don't provide any benefit there as of yet. So you may just be able to skip this step. For networking, usually provided buffers are a bit more interesting, on the receive side.

**Reply by jmillan** [reply]
Thanks @axboe, I'll verify the `ulimit -l` value when possible.

Regarding the usage of registered buffers with networking, the gain we are seeing is not big, around 2%, as detailed [here](https://github.com/axboe/liburing/discussions/1003#discussioncomment-7681488)[*] where flamegraphs for both cases (ZC and NO_ZC) can be compared. In summary it's basically `__ip_append_data()` the function that benefits from the registers. On the other side `udp_send_skb()`, which takes most of the CPU, does not seem to take any advantage, at least for the used network driver. 

 [*]: In order to properly render the .svg files, they may need to be downloaded to local host.

**Reply by axboe** [reply]
Gotcha, didn't think of the zero copy connection. Makes sense!

**Comment by isilence** [comment]
Note that buffers are rounded up to page size, when you register N 1500B buffers (i.e. struct iovec [N]) under the hood it'll account 4096 * N to the limit. It can be even more if some buffers span across multiple pages.
